src/catalog.rs

```diff
@@ -0,0 +1,126 @@
+use crate::result::IonResult;
+use crate::result::{illegal_operation, illegal_operation_raw};
+use crate::shared_symbol_table::SharedSymbolTable;
+use std::collections::{BTreeMap, HashMap};
+
+/// Catalog is a collection of Shared Symbol Tables
+/// for more information on [Catalog]: https://amzn.github.io/ion-docs/docs/symbols.html#the-catalog
+pub trait Catalog {
+    /// Returns the Shared Symbol Table with given table name
+    /// If a table with the given name doesn't exists or if the table name is an empty string
+    /// then returns an error
+    fn get_table(&self, name: &str) -> IonResult<SharedSymbolTable>;
+    /// Returns the Shared Symbol Table with given table name and version
+    /// If a table with given name and version doesn't exists then it returns an error
+    fn get_table_with_version(&self, name: &str, version: usize) -> IonResult<SharedSymbolTable>;
+}
+
+pub trait MutableCatalog {
+    /// Adds a Shared Symbol Table with name into the Catalog
+    fn put_table(&mut self, table: SharedSymbolTable);
+}
+
+struct SimpleCatalog {
+    tables_by_name: HashMap<String, BTreeMap<usize, SharedSymbolTable>>,
+}
+
+impl SimpleCatalog {
+    pub fn new() -> Self {
+        Self {
+            tables_by_name: HashMap::new(),
+        }
+    }
+}
+
+impl Catalog for SimpleCatalog {
+    fn get_table(&self, name: &str) -> IonResult<SharedSymbolTable> {
+        if name.is_empty() {
+            return illegal_operation("Can not get a symbol table where the name is empty");
+        }
+
+        let versions: &BTreeMap<usize, SharedSymbolTable> =
+            self.tables_by_name.get(name).ok_or_else(|| {
+                illegal_operation_raw(format!("Symbol table with name: {} does not exist", name))
+            })?;
+
+        let (_highest_version, table) = versions.iter().rev().next().ok_or_else(|| {
+            illegal_operation_raw(format!("Symbol table with name: {} does not exist", name))
+        })?;
+        Ok(table.to_owned())
+    }
+
+    fn get_table_with_version(&self, name: &str, version: usize) -> IonResult<SharedSymbolTable> {
+        if name.is_empty() {
+            return illegal_operation("Can not get a symbol table where the name is empty");
+        }
+
+        let versions: &BTreeMap<usize, SharedSymbolTable> =
+            self.tables_by_name.get(name).ok_or_else(|| {
+                illegal_operation_raw(format!("Symbol table with name: {} does not exist", name))
+            })?;
+
+        let table = versions.get(&version).ok_or_else(|| {
+            illegal_operation_raw(format!("Symbol table with name: {} does not exist", name))
+        })?;
+        Ok(table.to_owned())
+    }
+}
+
+impl MutableCatalog for SimpleCatalog {
```

Comment:
The `MutableCatalog` trait feels a bit off to me. Consider a few possible implementations of `Catalog`:
* `MapCatalog`: entirely in memory.
* `FileSystemCatalog`: looks for SST files with a given name within a given directory.
* `DynamoDbCatalog`: looks for SST documents in a Dynamo table.

`put_table` is a good fit for `MapCatalog` but doesn't fit as well for the other two. They could produce IO errors, and a user might wish to do bulk uploads of SSTs to a DynamoDB table rather than individual PUTs.

I propose turning this `impl MutableCatalog for SimpleCatalog` into just `impl SimplCatalog`; we can standardize a trait for reading from the `Catalog` while allowing each implementation to decide when and how to build the catalog up.

Reply:
Makes sense. I  will merge the `MutableCatalog` functionality into `SimpleCatalog` for this.